### PR TITLE
Add Top Level LPT tools as CLI

### PR DESCRIPTION
**Pull request by thomasdotred**
# Pull Request Checklist

- [ ] Read the [contributing guidelines](https://github.com/finbourne/lusid-python-tools/blob/master/docs/CONTRIBUTING.md)
- [ ] Tests pass

# Description of the PR

Describe the code changes for the reviewers, explain the solution you have provided and how it fixes the issue

### CODE REVIEW

File: lusidtools/commands/commands.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+
+def main():
+    tools = dict(find_tools())
+
+    args = sys.argv
+    if len(sys.argv) > 1:
+        first = sys.argv[1]
+        if first in tools:
+            import importlib
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: This import should be at the top of the file

File: lusidtools/commands/commands.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            mod = importlib.import_module(f"lusidtools.lpt.{tools[first][0]}")
+            sys.argv = [f'{sys.argv[0]} {first}'] + sys.argv[2:]
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: Can you please add a comment explaining what this is doing?

File: lusidtools/commands/commands.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import argparse
+import sys
+import lusidtools.lpt as lpt
+import os
+import ast
+
+
+def find_tools():
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: Please add a docstring

File: lusidtools/commands/commands.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                            if f.targets[0].id == "TOOLNAME":
+                                toolname = f.value.s
+                            elif f.targets[0].id == "TOOLTIP":
+                                tooltip = f.value.s
+                    if isinstance(f, ast.FunctionDef):
+                        if f.name == 'main':
+                            yield toolname, (toolmodule, tooltip)
+
+
+def help_method(args, tools):
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: Please add a docstring

File: lusidtools/commands/commands.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    def add_tip(tool, tip):
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: Please add a docstring

File: lusidtools/commands/commands.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import sys
+import lusidtools.lpt as lpt
+import os
+import ast
+
+
+def find_tools():
+    root = lpt.__path__[0]
+    for filename in os.listdir(root):
+        if filename.endswith('.py'):
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: To remove the nesting here we can do something like valid_files = [filename for filename in os.listdir(root) if filename.endswith('.py')]

Then

for file in valid files:
   etc.

File: lusidtools/commands/commands.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                        if (len(f.targets) == 1 and
+                                isinstance(f.targets[0], ast.Name) and
+                                isinstance(f.value, ast.Str)):
+                            if f.targets[0].id == "TOOLNAME":
+                                toolname = f.value.s
+                            elif f.targets[0].id == "TOOLTIP":
+                                tooltip = f.value.s
+                    if isinstance(f, ast.FunctionDef):
+                        if f.name == 'main':
+                            yield toolname, (toolmodule, tooltip)
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: This is fascinating. I've never seen this before. I understand what you are doing here and it is really nice. Can we limit the nesting to make it easier to understand? Anymore than 2 levels of nesting starts to get a little dicey and we are already a level deep here if you make the change above.

File: lusidtools/commands/commands.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    parser.add_argument('tool', metavar='<tool>', help='Tool to run. Options are listed below.')
+    parser.add_argument('arg1', metavar='<arg-1>', help=' - tool')
+    parser.add_argument('arg2', metavar=' ... ', help=' - specific')
+    parser.add_argument('arg3', metavar='<arg-n>', help=' - arguments')
+    parser.parse_args(['--help'])
+    exit()
+
+
+def main():
+    tools = dict(find_tools())
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: What is the shape of this dictionary given it comes from a generator?

File: lusidtools/commands/commands.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    args = sys.argv
+    if len(sys.argv) > 1:
+        first = sys.argv[1]
+        if first in tools:
+            import importlib
+            mod = importlib.import_module(f"lusidtools.lpt.{tools[first][0]}")
```
Comment (MikeMcGarry) [inline lusidtools/commands/commands.py]: Is tools[first][0] the toolmodule?

File: lusidtools/commands/commands.py (hunk 8 above)
Comment (thomasdotred) [reply lusidtools/commands/commands.py]:
```
{
    "qry_holdings": ("hld", "gets holdings of a portfolio")
}
```

File: lusidtools/commands/commands.py (hunk 9 above)
Comment (thomasdotred) [reply lusidtools/commands/commands.py]: yes, the name of the .py file

**Pull request merged**